cluster/src/main/scala/io/snappydata/gemxd/SparkSQLPrepareImpl.scala

```diff
@@ -0,0 +1,226 @@
+      DataSerializer.writeIntArray(types, hdos)
+    } else {
+      DataSerializer.writeIntArray(Array[Int](0), hdos)
+    }
+
+    if (msg.isLocallyExecuted) {
+      SparkSQLExecuteImpl.handleLocalExecution(srh, hdos)
+    }
+    msg.lastResult(srh)
+  }
+
+  override def serializeRows(out: DataOutput, hasMetadata: Boolean): Unit =
+    SparkSQLExecuteImpl.serializeRows(out, hasMetadata, hdos)
+
+  // Also see SnappyResultHolder.getNewNullDVD(
+  def getSQLType(dataType: DataType): (Int, Int, Int) = dataType match {
+    case IntegerType => (StoredFormatIds.SQL_INTEGER_ID, -1, -1)
+    case StringType => (StoredFormatIds.SQL_CLOB_ID, -1, -1)
+    case LongType => (StoredFormatIds.SQL_LONGINT_ID, -1, -1)
+    case TimestampType => (StoredFormatIds.SQL_TIMESTAMP_ID, -1, -1)
+    case DateType => (StoredFormatIds.SQL_DATE_ID, -1, -1)
+    case DoubleType => (StoredFormatIds.SQL_DOUBLE_ID, -1, -1)
+    case t: DecimalType => (StoredFormatIds.SQL_DECIMAL_ID,
+        t.precision, t.scale)
+    case FloatType => (StoredFormatIds.SQL_REAL_ID, -1, -1)
+    case BooleanType => (StoredFormatIds.SQL_BOOLEAN_ID, -1, -1)
+    case ShortType => (StoredFormatIds.SQL_SMALLINT_ID, -1, -1)
+    case ByteType => (StoredFormatIds.SQL_TINYINT_ID, -1, -1)
+    case BinaryType => (StoredFormatIds.SQL_BLOB_ID, -1, -1)
+    case _: ArrayType | _: MapType | _: StructType =>
+      // indicates complex types serialized as json strings
+      (StoredFormatIds.REF_TYPE_ID, -1, -1)
+
+    // send across rest as objects that will be displayed as json strings
+    case _ => (StoredFormatIds.REF_TYPE_ID, -1, -1)
+  }
+
+  def handleCase(branches: Seq[(Expression, Expression)], elseValue: Option[Expression],
+      datatype: DataType, nullable: Boolean, result: mutable.HashSet[ParamLiteral]): Unit = {
+    branches.foreach {
+      case (_, ParamLiteral(Row(pos: Int), NullType, 0)) =>
+        addParamLiteral(pos, datatype, nullable, result)
+    }
+    elseValue match {
+      case Some(ParamLiteral(Row(pos: Int), NullType, 0)) =>
+        addParamLiteral(pos, datatype, nullable, result)
+      case _ =>
+    }
+  }
+
+  def addParamLiteral(position: Int, datatype: DataType, nullable: Boolean,
+      result: mutable.HashSet[ParamLiteral]): Unit = if (!result.exists(_.pos == position)) {
+    result += ParamLiteral(nullable, datatype, position)
+  }
+
+  def allParamLiterals(plan: LogicalPlan, result: mutable.HashSet[ParamLiteral]): Unit = {
+    def allParams(plan: LogicalPlan): LogicalPlan = plan transformAllExpressions {
+      case bl@BinaryComparison(left: Expression, ParamLiteral(Row(pos: Int), NullType, 0)) =>
+        addParamLiteral(pos, left.dataType, left.nullable, result)
+        bl
+      case blc@BinaryComparison(left: Expression,
+      Cast(ParamLiteral(Row(pos: Int), NullType, 0), _)) =>
+        addParamLiteral(pos, left.dataType, left.nullable, result)
+        blc
+      case ble@BinaryComparison(left: Expression, CaseWhen(branches, elseValue)) =>
```

Comment:
I think this should be handled more generically instead of just CaseWhen. One should use the pattern used in other spark rules like

plan transform {

   case q: LogicalPlan => q transformAllExpressions {
    ....
}
}

Also, we should be able to inspect children of any expression and find out the paramliteral.